core/trino-main/src/main/java/io/trino/connector/WorkerDynamicCatalogManager.java

```diff
@@ -171,13 +170,14 @@ public void pruneCatalogs(Set<CatalogHandle> catalogsInUse)
             if (stopped) {
                 return;
             }
-            Iterator<Entry<CatalogHandle, Future<CatalogConnector>>> iterator = catalogs.entrySet().iterator();
-            while (iterator.hasNext()) {
-                Entry<CatalogHandle, Future<CatalogConnector>> entry = iterator.next();
-                if (!catalogsInUse.contains(entry.getKey())) {
-                    iterator.remove();
-                    futures.add(entry.getValue());
-                }
+
+            Set<CatalogHandle> catalogsToPrune = Sets.difference(catalogs.keySet(), catalogsInUse);
+            if (catalogsToPrune.isEmpty()) {
+                return;
+            }
+
+            for (CatalogHandle catalogHandleToPrune : catalogsToPrune) {
+                futures.add(catalogs.remove(catalogHandleToPrune));
```

Comment:
I'm not sure I like this change.  I can't put my finger on it but this reads as more risky because it is going throught the map multiple times.  I also dislike code that effectively does an is present check and then mutates the map.  Specifically the `remove` on the last line could return null (which would be an error) if there was a concurrent modification.  Obviously this code has a lock to prevent concurrent modification, which is why I'm not sure either way.  Anyway, I'm not saying that you should change this, but am saying found the original code more direct for me (that said I wrote the original, and I'm very old school java).